server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java

```diff
@@ -52,17 +64,102 @@ public Port getPort() {
     }
 
     public void start() {
+        ImmutableListMultimap<Endpoint, JMAPRoute> collect = jmapRoutes.stream()
+            .flatMap(JMAPRoutes::routes)
+            .collect(Guavate.toImmutableListMultimap(JMAPRoute::getEndpoint));
+
         if (configuration.isEnabled()) {
             server = Optional.of(HttpServer.create()
                 .port(configuration.getPort()
                     .map(Port::getValue)
                     .orElse(RANDOM_PORT))
-                .route(routes -> jmapRoutes.forEach(jmapRoute -> jmapRoute.define(routes)))
+                .route(routes -> jmapRoutes.forEach(jmapRoute -> collect.asMap().forEach(
+                    (endpoint, route) -> injectRoutes(routes, endpoint, route))))
                 .wiretap(configuration.wiretapEnabled())
                 .bindNow());
         }
     }
 
+    private HttpServerRoutes injectRoutes(HttpServerRoutes builder, Endpoint endpoint, Collection<JMAPRoute> routesList) {
+        if (routesList.size() == 1) {
+            JMAPRoute next = routesList.iterator().next();
+            switch (endpoint.getVerb()) {
```

Comment:
We can get rifd of this switch by making it a behaviour of the Verb class:

```
enum Verb {
   HttpServerRoutes registerRoute(String path, Action action) {...}
}
```

We can amlso add a similar method on the Endpoint object.

Then we can simply write:

```
return endpoint.register(builder, (req, res) ->getExistingRoute(extractRequestVersionHeader(req), next).apply(req, res));
```

Let's leverage object oriented capabilities ;-)

Reply:
This will allow getting rid of the duplicated `(req, res) ->                       getExistingRoute(extractRequestVersionHeader(req), next).apply(req, res));` code